Add flags for handling features of advanced logs

sairon opened the pull request:
Add flags for features previously available only to host logs.

mdegat01 commented on cmd/addons_logs.go:
```diff
@@ -54,12 +80,20 @@ Allowing you to look at the log output generated by a Home Assistant add-on.
+	addonsLogsCmd.Flags().BoolP("verbose", "v", false, "Return logs in verbose format")
+	addonsLogsCmd.Flags().Lookup("follow").NoOptDefVal = "true"
+	addonsLogsCmd.Flags().Lookup("verbose").NoOptDefVal = "true"
+
```
```suggestion

	addonsLogsCmd.RegisterFlagCompletionFunc("follow", boolCompletions)
	addonsLogsCmd.RegisterFlagCompletionFunc("lines", cobra.NoFileCompletions)
	addonsLogsCmd.RegisterFlagCompletionFunc("verbose", boolCompletions)
```
There should also be a `bootCompletions` ideally for the `boot` argument. That way users don't have to separately look up boot ids and copy and paste them. The completion would call `/host/logs/boots` and list out the boot IDs and offsets. See addon completions for an example of how this works:
https://github.com/home-assistant/cli/blob/713510d4333b0bb25d8d0b677d5369bd73861f7c/cmd/addons.go#L46-L101

mdegat01 reviewed (changes requested):
Need to add the completions as I suggested below.

Additionally I would strongly suggest refactoring this into two helpers: one for adding the flags and one for processing them. This same code is copied and pasted in 7 places it looks like, we should use common helpers instead.

sairon commented:
Thanks for the review, @mdegat01, it looks indeed better this way. There was some specific error handling implemented only for addons, not sure why. I have removed it, but if we want to keep it, I don't see a reason why only one of the endpoints should get special handling. Also added the completions, check the individual commits for less diff noise.

mdegat01 commented on cmd/root.go:
```diff
@@ -156,3 +158,57 @@ func initConfig() {
+
+func addLogsFlags(cmd *cobra.Command) {
+	cmd.Flags().BoolP("follow", "f", false, "Continuously print new log entries")
+	cmd.Flags().Int32P("lines", "n", 0, "Number of log entries to show")
```
```suggestion
	cmd.Flags().Uint32P("lines", "n", 0, "Number of log entries to show")
```
Should be uint right? Or do negative numbers make sense here? I forget if systemd-journal handles that

sairon replied:
That's actually copy-pasted from the old host logs, but right, I don't see a reason why it should be signed.

mdegat01 commented on cmd/host_logs_boots.go:
```diff
@@ -39,3 +40,31 @@ Show all values that can be used with the boot arg to find logs.
+		if boots, ok := data.Data["boots"].(map[string]interface{}); ok {
+			for bootID, bootName := range boots {
+				s := bootName.(string)
+				if toComplete == "" || strings.HasPrefix(s, toComplete) {
+					ret = append(ret, s)
+				}
+				if toComplete == "" || strings.HasPrefix(bootID, toComplete) {
+					ret = append(ret, bootID)
+				}
```
So this works. It seems a little odd to me that it'll go like this:
```
<boot hash>
0
<boot hash>
-1
...
```
Rather then this:
```
<boot hash>
<boot hash>
...
0
-1
...
```
But probably not worth the effort.

The only thing that might be worth the effort is you can add description to items. Like if you look at addon completions [here](https://github.com/home-assistant/cli/blob/713510d4333b0bb25d8d0b677d5369bd73861f7c/cmd/addons.go#L46). You can see how after the slug we provide the name and url for context. I think those boot hashes probably need the bootID prefix as context since just a random boot hash isn't particularly useful.

Maybe it would be better to just prompt bootIDs with the offsets as details or vice versa? I know then we aren't completing all the possible valid values but I think users using the completion really just to be able to pick any boot. Having 2 items per boot probably doesn't help them.

sairon replied:
Hmmm :thinking: At first I wanted to add just the hashes, as they are more annoying to write/paste. But then I realized that the offsets would be more common (in fact, I never thought of using the hashes, I always only used the offset) and having them completed makes the completion... complete :sweat_smile: 

I'll have a look how to improve the UX of that, but if we should pick only one, then I'll hide the offsets and complete only hashes. But having both side-by-side in completion also has something to it...

In regard to:

> since just a random boot hash isn't particularly useful.

I consider the completion as a tool to make the task of writing something less annoying - so when I remember that the hash went like `68ed...`, I type `6` and then tab and add more tokens until I get the right value. So I disagree here - completing a random hash still might be useful.

sairon replied:
Agreed. With the descriptions (didn't know about them before) it also makes it even more pointless to complete it. The completion looks like this now:
```
db32e85e327a49bcbd05f4f3ea09eedf	boot offset -3
b3ed8ad3f86840778910e816cdad7167	boot offset -2
22100f35cd594bf78d0d574cb403feb1	boot offset -1
9abe1d6d7b004134993e40a8e37d0e36	boot offset 0
73cd933be6164cada777f4f48f5dc10a	boot offset -7
2b76a0fb7e354f859ed9b88ff86f2865	boot offset -6
84752b1466be48608b3ef4105c923b09	boot offset -5
4beae2e05fac4b3893a3649833d32f64	boot offset -4
```

The order changes every time, because Go doesn't guarantee order of iteration over maps. Are we fine with that?

agners replied:
Ideally I guess it would be ordered by boot offset, but then, doesn't matter all that much to me :man_shrugging:

agners replied:
Yeah it seems a pain to change that. Let's go with what we have now.

mdegat01 reviewed (approved):
This looks good.

I guess my minor organization suggestion would be I think it makes more sense to have `addLogsFlags` and `processLogsFlags` in `cmd/host.go` rather then `cmd/root.go`. I would also probably put `hostBootCompletions` in there for good measure. But not necessary, it all works fine like this.

sairon commented on cmd/host_logs_boots.go:
```diff
@@ -39,3 +40,31 @@ Show all values that can be used with the boot arg to find logs.
+		if boots, ok := data.Data["boots"].(map[string]interface{}); ok {
+			for bootID, bootName := range boots {
+				s := bootName.(string)
+				if toComplete == "" || strings.HasPrefix(s, toComplete) {
```
Haha, thanks! That's purely coincidental, thanks goes also to syntax highlighting in GoLand, as the greyed unused parameter caught my attention there. And it's also vital for my shell completion brute-forcing I described above :sweat_smile:
